applications/sarp-2-swdesign.md

```diff
@@ -0,0 +1,155 @@
+
+- https://github.com/masapr
+- https://github.com/haerdib
+- https://github.com/echevrier
+- https://github.com/Niederb
+
+
+### Team LinkedIn Profiles
+
+- https://www.linkedin.com/in/sabine-proll-5a7118153
+- https://www.linkedin.com/in/bigna-h%C3%A4rdi-736bb21a9
+- https://www.linkedin.com/in/edith-chevrier-90233297
+- https://www.linkedin.com/in/thomas-niederberger-6057b71a7
+
+## Development Status :open_book:
+
+In a first research project we investigated, if MIRAI can be used for static code analysis of substrate pallets. For this we did a proof of concept on two cases: 
+- Check of [incorrect origin](https://github.com/scs/MIRAI/blob/Milestone1_Research/substrate-examples/pallet_template/README.md) in the [substrate node template](https://github.com/substrate-developer-hub/substrate-node-template/tree/e0c480c0f322d0b0d1b310c93fa646fc0cfdd2df/pallets/template)
+- Validation of [unsigned transactions](https://github.com/scs/MIRAI/blob/Milestone1_Research/substrate-examples/offchain-worker/README.md) for substrate's [offchain worker example](https://github.com/paritytech/substrate/tree/ea9ce4c0af36310c0b0db264ab12cf4766b83750/frame/examples/offchain-worker)
+
+The overall conclusion was, that it is best to run the analysis only on the newly written pallet code, but not on the code generated by substrate's macros. To facilitate this a detailed analysis of different software designs has to be evaluated.
+
+
+The full documentation of our findings can be found [here](https://github.com/scs/MIRAI/tree/Milestone1_Research/substrate-examples).
+
+
+## Development Roadmap :nut_and_bolt:
+
+
+
+### Overview
+
+- **Total Estimated Duration:** 1 month
+- **Full-Time Equivalent (FTE):** 1 FTE
+- **Total Costs:** 30,000 USD
+
+
+### Milestone 1 Software Design & Bug Fixes
+
+- **Estimated duration:** 1 month
+- **FTE:**  1
+- **Costs:** 30,000 USD
+
+In our previous work, we found the following problems:
+
+1. **Crashes and timeouts of MIRAI** Certain pieces of substrate code lead to crashes of MIRAI. In other cases, parts of the code are not analyzed/do not produce warnings, because MIRAI runs into a timeout before reaching this code. Because of this, our examples are rather simple and we couldn't add and check tags at the locations we originally wanted to.
+
+2. **Complexity due to substrate macros** The main reason for crashes and timeouts in our examples, was caused by substrate macros, adding a lot of complexity to the code in the background. Ideally SARP only analyzes the newly written code of a pallet.
+
+3. **Invasiveness of tag analysis** The code we wrote in our PoC is very invasive and changes the code of the pallet. This is not practical for end-users. Ideally the user doesn't need to change anything on their code, or at least the changes should be very simple.
+
+To address 2. and 3. we plan to evaluate different software designs. These will be part of our deliverables and we plan to discuss these with Parity and/or Web3 Foundation.

```

Comment:
An interesting approach could be to extract the MIR from MIRAI, abstract out low-level details and just track the flow of FRAME level function calls. This is something which will require tweaking in to MIRAI code, and reusing their sink-source analysis engine on a more abstract MIR. 
Please add more details on other designs being considered.